Add covariance

efernandez opened the pull request:
Add pose and twist covariance.

Also add tests for the covariance and the twist fixed in https://github.com/clearpathrobotics/ros_controllers/pull/2.

Also roslint has been added and several warnings fixed.

Commit:
Add covariance

Commit:
Support different left and right wheel radii

This is needed to compute the covariance properly

Commit:
Add propagation test

Check that the following propagation equations are satisfied:

(x, y) = (x, y) + R(yaw) * (v_x, v_y) * dt
yaw = yaw + v_yaw * dt (normalized)

Commit:
Add roslint and fix several warnings

Commit:
Add covariance tests

1. Moving along x axis covariance grows faster on y axis
2. Moving along y axis covariance grows faster on x axes

efernandez commented:
@servos @paulbovbel 

I'll squash the commits after your review.

For the positive definite test/check on the covariance matrix I cannot compared against 0 because of numerical issues with Eigen eigensolver for self-adjoint matrices. I've also tried the Cholesky decomposition, and it's only slightly more robuts. With Matlab `eig` the test passes. As a workaround, which I think it's appropriate in this case, I compared against `-EPS`. With this, all the tests pass.

servos commented on diff_drive_controller/include/diff_drive_controller/direct_kinematics_integrate_functor.h:
```diff
@@ -0,0 +1,111 @@
+      const T vl = v_l * T(left_wheel_radius_);
+      const T vr = v_r * T(right_wheel_radius_);
+
+      const T v = (vr + vl) * T(0.5);
```
This fails (ie vl = 0) if T is an int (not that that is likely). Any reason not to do (vr + vl) / T(2)?

efernandez replied:
`T` must be double/float or a dual number like `Jet<double/float, N>`. It should never be `int`, but I'm not sure how to avoid that.

I use `T(0.5)` as it was done before, no difference with dividing by `T(2)`.

servos replied:
Fair enough.

mikepurvis replied:
Depending how strongly you feel, you can use `static_assert` to prevent the template being instantiated with inappropriate types, eg:

```
#include <boost/type_traits.hpp>
static_assert(!boost::is_floating_point<T>::value, "The pose components must be specified as float values.");
```

(Note that the type_traits module has been brought into the STL for c++11.)

efernandez replied:
Good one.

Since C++11 is not allowed upstream, I've used `BOOST_ASSERT_MSG` instead of `static_assert` tough.

paulbovbel replied:
I think you're looking for BOOST_STATIC_ASSERT_MSG for compile time

efernandez replied:
Yes, that's true. Fixed.

It seems that with Boost the message isn't shown, and instead a quite strange compilation error is shown. This doesn't happen with `static_assert` (C++11) tough. See the two links for comparison purposes [here](http://stackoverflow.com/a/10585543/2399619).

I've decided to keep the check (using Boost, since C++11 isn't recommended in ROS as I understand from [REP3](http://www.ros.org/reps/rep-0003.html#c)), since the case when this happens is uncommon and at least the error points to the line with the static assert.

servos commented on diff_drive_controller/src/diff_drive_controller.cpp:
```diff
@@ -511,33 +586,39 @@ namespace diff_drive_controller{
+
     odom_pub_->msg_.pose.pose.position.z = 0;
-    odom_pub_->msg_.pose.covariance = boost::assign::list_of
-        (static_cast<double>(pose_cov_list[0])) (0)  (0)  (0)  (0)  (0)
```
Should we change the param name of these covariance lists since they no longer mean what they used to?

efernandez replied:
Yes, or even remove it, since an initial value should be a small thing.

In the code now I've hardcoded initial values for the diagonal, with `1e-6` and `1e-9` for the pose and twist covariance matrices, respectively. But I think this are needed any way, because they are used at the constructor.

Actually, we could change those params to something like initial_\* and only take the 3 important elements, `x`, `y` and `\theta`.

efernandez replied:
Done.

Now there's only `initial_pose_covariance_diagonal` with 3 elements: `x`, `y`, `\theta` (or yaw).
And it's optional.

Note that twist covariance is computed every cycle, so any value from the initialization would be discarded.

The pose covariance might still be initialized to something that the user considers appropriate, but by default it takes small values just to avoid non-positive definite and ill-conditioned matrices.

efernandez commented:
BTW, after `make test` all tests pass:

```
100% tests passed, 0 tests failed out of 12
```

Now there are many more tests, in the sense that for the basic controller test also the covariance is tested.

efernandez commented on diff_drive_controller/include/diff_drive_controller/autodiff_integrate_function.h:
```diff
@@ -0,0 +1,145 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2013, PAL Robotics, S.L.
```
@mikepurvis Should I use a Clearpath Robotics copyright for the new files I've created here? If that's the case, could you provide me a link to an example?

mikepurvis replied:
https://github.com/clearpathrobotics/headers/blob/master/bsd/cpp

efernandez replied:
Thanks.

I'll make it 80-column width tough.

efernandez commented:
@paulbovbel Any comments before I start squashing commits and merging?

efernandez commented on diff_drive_controller/package.xml:
```diff
@@ -14,12 +14,16 @@
   <build_depend>realtime_tools</build_depend>
   <build_depend>tf</build_depend>
   <build_depend>urdf</build_depend>
+  <build_depend>eigen</build_depend>
```
@mikepurvis Do you know if this is actually needed when cmake_modules is already a build depend, following [this](https://github.com/ros/cmake_modules/blob/0.3-devel/README.md#usage)?

mikepurvis replied:
I believe so. `cmake_modules` only gets you the ability to `find_package` it, but you still need the package itself to be present on the system.

efernandez commented:
I've just squashed the commits.
I've also added a comment to make it clear that a covariance matrix is positive semidefinite, but we check for positive definite (up to some EPS precision) because in the tests we know (or want to enforce that) the covariance matrices must be positive definite.

If there are no additional comments, I'll merge this.
